tests/test_router_quote_in_given_out.py

```diff
@@ -0,0 +1,243 @@
+import time
+import pytest
+
+import conftest
+from sqs_service import *
+from quote import *
+from quote_response import *
+from rand_util import *
+from e2e_math import *
+from decimal import *
+from constants import *
+from util import *
+
+ROUTES_URL = "/router/quote"
+
+QUOTE_NUM_TOP_LIQUIDITY_DENOMS = 20
+
+# Arbitrary choice based on performance at the time of test writing
+EXPECTED_LATENCY_UPPER_BOUND_MS = 15000
+
+# The max amount in value in USD to run the price impact check
+# This is primarily to avoid flakiness due to swapping large amounts.
+# The choice is arbitrary and was made based on testing at the time of creation.
+# In the future, we might lower or increase this value based on the performance of the system.
+HIGH_LIQ_PRICE_IMPACT_CHECK_USD_AMOUNT_IN_THRESHOLD = 5000
+
+# The max price impact threshold for the high liquidity check
+HIGH_LIQ_MAX_PRICE_IMPACT_THRESHOLD = 0.5
+
+# Test suite for the /router/quote endpoint
+# Test runs tests for exact amount out quotes.
+class TestExactAmountOutQuote:
+    @pytest.mark.parametrize("coin_obj", construct_token_in_combos(conftest.choose_tokens_liq_range(QUOTE_NUM_TOP_LIQUIDITY_DENOMS), USDC_PRECISION - 1, USDC_PRECISION + 4), ids=id_from_coin)
+    def test_usdc_in_high_liq_in(self, environment_url, coin_obj):
+        """
+        This test case validates quotes between USDC in and NUM_TOP_LIQUIDITY_DENOMS.
+        The amounts are constructed to be seeded random values between 10^USDC_PRECISION-1 and 10 ^(USDC_PRECISION + 4)
+
+        This allows us to validate that we can continue to quote at reasonable USDC values for all major token pairs without errors.
+
+        Note: the reason we use Decimal in this test is because floats truncate in some edge cases, leading
+        to flakiness.
+        """
+
+        denom_in = coin_obj["denom"]
+        amount_str = coin_obj["amount_str"]
+        amount_out = int(amount_str)
+
+        # Choose the error tolerance based on amount in swapped.
+        error_tolerance = Quote.choose_error_tolerance(amount_out)
+
+        # Skip USDC quotes
+        if denom_in == USDC:
+            return
+
+        denom_in_data = conftest.shared_test_state.chain_denom_to_data_map.get(denom_in)
+        denom_in_precision = denom_in_data.get("exponent")
+        
+        # Compute spot price scaling factor.
+        spot_price_scaling_factor = Decimal(10)**6 / Decimal(10)**denom_in_precision
+
+        # Compute expected spot prices
+        out_base_in_quote_price = Decimal(denom_in_data.get("price"))
+        expected_out_base_in_quote_price = 1 / out_base_in_quote_price
```

Comment:
Should `expected_out_base_in_quote_price` be `expected_in_base_out_quote_price` because we inverted the original `out_base_in_quote_price`?

Reply:
Good catch, updated! :+1: